designs/2024-config-lookup-from-file/README.md

```diff
@@ -0,0 +1,181 @@
+1. **More flat config changes.** With some of the criticisms around flat config, we may get some blowback for making another change while most of the ecosystem hasn't yet converted. Hopefully, though, this change will have minimal impact on most users.
+
+## Backwards Compatibility Analysis
+
+While a breaking change, this proposal will have a minimal impact on most users of flat config. Because linted files typically exist under the current working directory, and the configuration file is either in the current working directory or an ancestor directory, projects that currently use a flat config file will likely not see any difference in the way ESLint works. The same config file will be found during the resolution process.
+
+This is breaking primarily because users might have other config files deeper in the project directory hierarchy that might be found instead of the one found by searching from the cwd.
+
+## Alternatives
+
+* **Lookup strategy CLI/`ESLint` Option.** Instead of migrating towards a new config lookup strategy, we could also let people opt in via a command line argument and associated option to the `ESLint` constructor. Something like `--config-lookup-from file` with the default being `--config-lookup-from cwd`. The downsides of this approach are that it requires people to opt-in (at least until we can switch the default) and we'll be stuck maintaining two different config lookup strategies.
+* **Configurable lookup strategy.** We could also allow people to specify a lookup strategy, or even a custom implementation of a lookup strategy, using a configuration file (either in `eslint.config.js` or as a distinct file). This would be more complicated to implement and maintain.
+
+## Open Questions
+
+N/A
+
+## Help Needed
+
+N/A
+
+## Frequently Asked Questions
+
+### What will the behavior be of `-c` when `*_config_lookup_from_file` is enabled?
+
+The same as the current implementation: The `-c` option completely overrides which config file to use and there will be no config lookup performed.
+
+### If I'm linting `./subdir/foo.js` where `./subdir/eslint.config.js` exists, why is `./eslint.config.js` loaded?
+
+Here's the structure in question:
+
+```
+/usr/tmp/
+├── eslint.config.js
+└── subdir/
+    ├── foo.js
+    └── eslint.config.js
+```
+
+Because ignores are handled in `eslint.config.js`, we need to look at `./eslint.config.js` to ensure that `subdir` isn't being ignored.
+
+### What happens if there's a `eslint.config.js` file with an ignore pattern of `subdir` in the parent directory `subdir`?
+
+Here's the structure in question:
+
+```
+/usr/tmp/
+├── eslint.config.js    <-- ignores: ["subdir"]
+└── subdir/
+    ├── foo.js
+    └── eslint.config.js
+```
+
+When `eslint subdir/foo.js` is called, no files will be linted because the root `eslint.config.js` file ignores `subdir`.
```

Comment:
With `issue18385c` branch, in this scenario, `eslint subdir/foo.js` does lint the file. `eslint subdir` doesn't though.

Reply:
Also, if linting `subdir` is run from another subdirectory, it does lint `subdir`, which looks a bit inconsistent.

```shell
$ npx eslint --flag unstable_config_lookup_from_file subdir/foo.js

C:\projects\tmp\tmp\subdir\foo.js
  1:1  error  'a' is not defined  no-undef

✖ 1 problem (1 error, 0 warnings)


milos@DESKTOP-4GFK8EQ MINGW64 /c/projects/tmp/tmp
$ npx eslint --flag unstable_config_lookup_from_file subdir

Oops! Something went wrong! :(

ESLint: 9.5.0

You are linting "subdir", but all of the files matching the glob pattern "subdir" are ignored.

If you don't want to lint these files, remove the pattern "subdir" from the list of arguments passed to ESLint.

If you do want to lint these files, explicitly list one or more of the files from this glob that you'd like to lint to see more details about why they are ignored.

  * If the file is ignored because of a matching ignore pattern, check global ignores in your config file.
    https://eslint.org/docs/latest/use/configure/ignore

  * If the file is ignored because no matching configuration was supplied, check file patterns in your config file.
    https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-with-arbitrary-extensions

  * If the file is ignored because it is located outside of the base path, change the location of your config file to be in a parent directory.


milos@DESKTOP-4GFK8EQ MINGW64 /c/projects/tmp/tmp
$ cd subdir1

milos@DESKTOP-4GFK8EQ MINGW64 /c/projects/tmp/tmp/subdir1
$ npx eslint ../subdir

C:\projects\tmp\tmp\subdir\foo.js
  1:1  error  'a' is not defined  no-undef

✖ 1 problem (1 error, 0 warnings)
```